KAFKA-16606 Gate JBOD configuration on 3.7-IV2

soarez opened the pull request:
Support for multiple log directories in KRaft exists from MetataVersion 3.7-IV2.

When migrating a ZK broker to KRaft, we already check that the IBP is high enough before allowing the broker to startup.
But we still allow a KRaft broker to be configured with multiple directories regardless of the metadata version,
and whether JBOD is supported or not.
Running  a KRaft broker with multiple log directories before 3.7-IV2 means the Controller will not re-elect new
leaders when a directory fails, leading to indefinite unavailability.

With KIP-584 and KIP-778, Brokers in KRaft mode do not require the IBP configuration - the configuration is deprecated. In KRaft mode inter.broker.protocol.version defaults to MetadataVersion.MINIMUM_KRAFT_VERSION (IBP_3_0_IV1).

Instead KRaft brokers discover the MetadataVersion by reading the "metadata.version" FeatureLevelRecord from the cluster metadata.

This change adds a new configuration validation step upon discovering the "metadata.version" from the cluster metadata.

cc @cmccabe since this seems related to [KIP-920](https://cwiki.apache.org/confluence/x/pAo0Dw)

soarez committed:
KAFKA-16606 Gate JBOD configuration on 3.7-IV2

Support for multiple log directories in KRaft exists from
MetataVersion 3.7-IV2.

When migrating a ZK broker to KRaft, we already check that
the IBP is high enough before allowing the broker to startup.

With KIP-584 and KIP-778, Brokers in KRaft mode do not require
the IBP configuration - the configuration is deprecated.
In KRaft mode inter.broker.protocol.version defaults to
MetadataVersion.MINIMUM_KRAFT_VERSION (IBP_3_0_IV1).

Instead KRaft brokers discover the MetadataVersion by reading
the "metadata.version" FeatureLevelRecord from the cluster metadata.

This change adds a new configuration validation step upon discovering
the "metadata.version" from the cluster metadata.

mimaison commented:
Thanks for the PR.

If I set `inter.broker.protocol.version=3.6-IV2` and multiple log directories in my broker configuration, it now prevents me from formatting the storage:
```
bin/kafka-storage.sh format -t $KAFKA_CLUSTER_ID -c config/kraft/server.properties
[2024-05-15 17:05:31,009] WARN inter.broker.protocol.version is deprecated in KRaft mode as of 3.3 and will only be read when first upgrading from a KRaft prior to 3.3. See kafka-storage.sh help for details on setting the metadata.version for a new KRaft cluster. (kafka.server.KafkaConfig)
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: Multiple log directories (aka JBOD) are not supported in the configured inter.broker.protocol.version 3.0-IV1. Need 3.7-IV2 or higher
```

However, if I specify the metadata version using the `--release-version` flag of the storage tool, the formatting works and I only get the failure when I try to start the broker:
```
bin/kafka-storage.sh format -t $KAFKA_CLUSTER_ID -c config/kraft/server.properties --release-version 3.6
metaPropertiesEnsemble=MetaPropertiesEnsemble(metadataLogDir=Optional.empty, dirs={/tmp/kraft-combined-logs: EMPTY, /tmp/kraft-combined-logs2: EMPTY})
Formatting /tmp/kraft-combined-logs with metadata.version 3.6-IV2.
Formatting /tmp/kraft-combined-logs with metadata.version 3.6-IV2.
Formatting /tmp/kraft-combined-logs2 with metadata.version 3.6-IV2.
```

Should we ensure the failure always happens at the formatting stage?

I also created https://issues.apache.org/jira/browse/KAFKA-16771 to avoid printing the "Formatting <...>" log line twice for the first directory.

soarez committed:
format command validates config for given MV

soarez commented:
@mimaison thanks for having a look.

Yes, I think we should check the MetadataVersion at the formatting stage too. Please have a look at my latest changes.

soarez commented:
@mimaison Sorry, thanks for pointing that out. That was an unintentional, bad rebase. Sorted now.

mimaison commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1360,6 +1360,14 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      && logDirs.size > 1) {
+        require(interBrokerProtocolVersion.isDirectoryAssignmentSupported,
+          s"Multiple log directories (aka JBOD) are not supported in the configured " +
+            s"${ReplicationConfigs.INTER_BROKER_PROTOCOL_VERSION_CONFIG} ${interBrokerProtocolVersion}. " +
```
Should we add `ReplicationConfigs.INTER_BROKER_PROTOCOL_VERSION_CONFIG` after the value? So the message looks like `Multiple log directories (aka JBOD) are not supported with the configured 3.6 inter.broker.protocol.version`

mimaison commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1457,6 +1465,18 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+   * Validate some configurations for new MetadataVersion. A new MetadataVersion can take place when
+   * a FeatureLevelRecord for "metadata.version" is read from the cluster metadata.
+   */
+  def validateWithMetadataVersion(metadataVersion: MetadataVersion): Unit = {
```
This seems to be the exact same check we do in `validateValues()`. Can we consolidate both?

soarez replied:
We can call this from `validateValues`, but then I won't be able to take your other suggestion to mention `inter.broker.protocol.version` in the error message. This one gets called when the KRaft broker discovers a new metadata.version via a new feature record. Would you prefer we consolidate the checks, or keep mention of `inter.broker.protocol.version` when it makes sense? 
I lean towards leaving them both, so it's helpful for the operator, and the duplication cost should go away in 4.0.

mimaison replied:
Ah right, thanks for the explanation. Then I guess it makes sense to keep both methods for now.

mimaison reviewed (commented):
Thanks for the update! I left a couple of comments but did not check the StorageTool changes yet since it needs rebasing.

soarez committed:
Improve IBP validation message

mimaison commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -45,89 +45,124 @@ import scala.jdk.CollectionConverters._
+  def runMain(args: Array[String]): Int = {
+    val namespace = parseArguments(args)
+    val command = namespace.getString("command")
+    val config = parseConfig(namespace.getString("config"))
```
We can't run this when the command is `random-uuid`, as `KafkaConfig` requires some configs to be set and we don't provide any:
```
KAFKA_CLUSTER_ID="$(bin/kafka-storage.sh random-uuid)"
Exception in thread "main" org.apache.kafka.common.config.ConfigException: Missing required configuration `zookeeper.connect` which has no default value.
	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:1223)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1214)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:544)
	at kafka.tools.StorageTool$.parseConfig(StorageTool.scala:91)
	at kafka.tools.StorageTool$.runMain(StorageTool.scala:57)
	at kafka.tools.StorageTool$.main(StorageTool.scala:79)
	at kafka.tools.StorageTool.main(StorageTool.scala)
```

soarez replied:
You're right, thanks for pointing this out. I missed that when I simplified the `Option(...).flatMap()` config bit. I've brought that back now.

mimaison commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -45,89 +45,124 @@ import scala.jdk.CollectionConverters._
+    try {
+      new KafkaConfig(Utils.loadProps(configFilename))
+    } catch {
+      case e: IllegalArgumentException => throw new TerseFailure(s"Invalid configuration: ${e.getMessage}")
```
Wrapping `IllegalArgumentException` slightly changes the behavior in case of bad configurations. For example if I set `log.dirs=`, on trunk I get:
```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: At least one log directory must be defined via log.dirs or log.dir.
	at scala.Predef$.require(Predef.scala:337)
	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2316)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:2285)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1634)
	at kafka.tools.StorageTool$.$anonfun$main$1(StorageTool.scala:52)
	at scala.Option.flatMap(Option.scala:283)
	at kafka.tools.StorageTool$.main(StorageTool.scala:52)
	at kafka.tools.StorageTool.main(StorageTool.scala)
```

With this PR I get:
```
Invalid configuration: requirement failed: At least one log directory must be defined via log.dirs or log.dir.
```

soarez replied:
I reverted this with the config parsing issue mentioned above.

mimaison reviewed (commented):
Thanks for the update, I made another pass and left a few questions.

soarez committed:
Lazily evaluate config as necessary

soarez committed:
Use dedicated publisher

soarez commented:
@mimaison I rebased due to conflicts and changed how the MV change via metadata records is detected to use a publisher following #15945. Please take another look, thanks.

mimaison commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -45,89 +45,117 @@ import scala.jdk.CollectionConverters._
-        case _ =>
-          throw new RuntimeException(s"Unknown command $command")
-      }
+  def main(args: Array[String]): Unit = {
```
Can we move `main()` to the top?

mimaison commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -45,89 +45,117 @@ import scala.jdk.CollectionConverters._
+   * @param args The command line arguments
+   * @return     The exit code
+   */
+  def runMain(args: Array[String]): Int = {
```
In other tools, the method that is called by `main()` is typically called `execute()`. I think it would make sense to use that name here too.

soarez replied:
I hadn't noticed this yet. Thanks for pointing this out.

mimaison commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -656,5 +657,47 @@ Found problem:
+    def formatWith(logDirCount: Int, metadataVersion: MetadataVersion): Integer = {
+      val properties = TestUtils.createBrokerConfig(10, null, logDirCount = logDirCount)
+      properties.remove(ReplicationConfigs.INTER_BROKER_PROTOCOL_VERSION_CONFIG)
+      StorageTool.runMain(Array("format",
```
We can get rid of `createPropsFile()` and reuse `TestUtils.tempPropertiesFile()`
```suggestion
val configFile = TestUtils.tempPropertiesFile(properties.asInstanceOf[util.Map[String, String]].asScala)
StorageTool.runMain(Array("format",
        "-c", configFile.getAbsolutePath,
        "-t", "XcZZOzUqS4yHOjhMQB6JLQ",
        "--release-version", metadataVersion.toString))
```

soarez replied:
Makes sense

mimaison commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -656,5 +657,47 @@ Found problem:
+  def testJbodSupportValidation(): Unit = {
+    def formatWith(logDirCount: Int, metadataVersion: MetadataVersion): Integer = {
+      val properties = TestUtils.createBrokerConfig(10, null, logDirCount = logDirCount)
+      properties.remove(ReplicationConfigs.INTER_BROKER_PROTOCOL_VERSION_CONFIG)
```
It looks like we're missing a test that check when inter broker protocol version is set?

soarez replied:
This property is ignored in KRaft. It assumes the value of `MINIMUM_KRAFT_VERSION` and cannot be set.

https://github.com/apache/kafka/blob/f2aafcc66faca4d07a27e4ef90158136cb317b44/core/src/main/scala/kafka/server/KafkaConfig.scala#L896-L914

So I don't think we need to test this? What test did you have in mind?

mimaison replied:
Ah right, thanks for checking.

mimaison commented on core/src/main/java/kafka/server/MetadataVersionConfigValidator.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.kafka.server.common.MetadataVersion;
+import org.apache.kafka.server.fault.FaultHandler;
+
+public class MetadataVersionConfigValidator implements MetadataPublisher {
```
Sorry I'm not super familiar yet with this. Can you explain why we would catch the configuration error here instead of in KafkaConfig?

soarez replied:
I'm glad you asked.

There are two ways the broker can assume a MetadataVersion:

1. Via the `inter.broker.protocol.version` configuration property in ZK mode
2. Via discovering a `FeatureRecord` for `metadata.version` in the cluster metadata in KRaft mode. This is preset with `--release-version` in the format command or with `kafka-features.sh upgrade --metadata`.

For (1) we have the version set at startup time, and we can catch the conifguration error in `KafkaConfig` as the properties are loaded. But for (2) we can only validate config later when the version is discovered from metadata.

Ealier in this PR I had the check in `BrokerMetadataPublisher` — this class streams metadata records in the broker and distributes them to various types of `MetadataPublisher` for reconciliation — but instaed of adding the check to that class I think it's better reuse the publisher pattern, similar to how the broker registration re-sending was extracted to `BrokerRegistrationTracker` in #15945.

Perhaps we could implement `MetadataPublisher` directly in `KafkaConfig`, but I thought that class is big enough already.

Let me know what you think.

soarez committed:
Move main() to top

soarez committed:
Rename runMain() to execute()

soarez committed:
Reuse TestUtils.tempPropertiesFile()

mimaison reviewed (approved):
LGTM, thanks for the fix!

mimaison commented:
You can backport it to 3.7 too if you want.

soarez commented:
Thanks for reviewing this @mimaison. Backporting to 3.8 and 3.7.

ableegoldman commented:
Hey @soarez / @mimaison -- I've noticed the ReassignPartitionsCommandTest.testReassignmentCompletionDuringPartialUpgrade test seems to be failing consistently since this was merged, with an error message that seems to be related to my unfamiliar eyes. Can you take a quick look into this failing test?

You can find the full logs on the [build for this PR](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15834/7/tests), but here is the fatal exception:

```
java.lang.IllegalArgumentException: requirement failed: Multiple log directories (aka JBOD) are not supported with the configured 3.0-IV1 inter.broker.protocol.version. Need 3.7-IV2 or higher
```

soarez commented:
@ableegoldman @mimaison please see  #16251
